### DISCUSSION TITLE
Use of existing C++ libraries (like wxWidget or other C or Objective-C libraries)

### DISCUSSION THREAD

**Discussion by sparky4pro** [julelang/jule#26] [Discussions]
Do I need to create **C++** 'glue code' for existing **C++** libraries (like **wxWidget**) or can I use them with normal **Jule** code?
Any hints how to use existing **C++** (or **C** or **Objective-C**) libraries are welcome.
Thank you in advance.

**Comment by mertcandav** [comment]
In theory, it is possible to use existing libraries. However, it is often a valid statement that it is not possible to do this directly. As can be seen in the [C++ interoperability documentation](https://jule.dev/pages/manual.html?page=cpp-interoperability), a binding is required. But if the types are compatible, only binding is sufficient. You don't need to do anything additional in any way.

In addition, Jule does not have all the design elements of C++. The most common situation that can cause problems when using an external library is when the external library requires a namespace. Therefore, you may need to write a custom binding header file for Jule and define functions that allow you to avoid namespaces.

For example:
```cpp
jule::Int get_magic_number() {
    return mylibrary::get_magic_number();
}
```
The code above is a binding-compatible C++ function wrapper for the `get_magic_number` function in the `mylibrary` namespace.

As for what kind of bindings Jule supports, the [C++ interoperability documentation](https://jule.dev/pages/manual.html?page=cpp-interoperability) mentions all types of bindings. Anything other than these is not supported. If you need a wrapper feel free to make it, you probably won't have a problem. However, if you want to create a fully compatible Jule binding of a 3rd party C++ library, you may need to spend some time doing this, but it will probably take much less time than writing the entire library in Jule.

**Comment by sparky4pro** [comment]
Thank you very much for your thorough explanation.
Will post any success or additional questions here.

**Comment by sparky4pro** [comment]
From the interoperability documentation:
```v
// sum.hpp
using namespace jule;

Int sum(const Slice slice) {
    Int total{ 0 };
    for (const Int x: slice)
        total += x;
    return total;
}
```
```v
// main.jule
use cpp "sum.hpp"

cpp fn sum([]int): int

fn main() {
    let numbers = [1, 2, 3, 4, 5, 6, 7, 8]
    let total = cpp.sum(numbers)
    outln(total)
}
```
does not compile, and the only message is: `exit status 1`.

Using clang separate I get the following error message:
```
ir.cpp:7:
sum.hpp:3:15: error: use of class template 'Slice' requires template arguments
Int sum(const Slice slice) {
              ^
slice.hpp:24:11: note: template is declared here
    class Slice {
          ^
1 error generated.
```
Would be nice if juleC would show clang's error messages.

**Reply by mertcandav** [reply]
Sorry. The source code is written in wrong HTML, so the generic type is not given. That's why the code is can't compiled. The fix has been published for website.

Your suggestion for JuleC to log compiler output will be taken into accounts. Obviously this has been considered before but has not been added. It will be evaluated again.  In the worst case, a notice will be added as to why the compiler might not be compiled.

Thanks.

**Comment by sparky4pro** [comment]
Yes, this was it. Thank you.

**Comment by sparky4pro** [comment]
Adding normal `*.c` and `*.h` files works with no issues.

Using `Objective-C` doesn't, as the compiler doesn't recognize even simple types like a `NSString` and it seems `clang` doesn't switch on certain things in order to handle `Objective-C` files.

Digging through the internet revealed that `C++` code cannot be used with `Objective-C` code without renaming `*.cpp` files to `*.mm` while doing a manual compile, which I tried.
And the outcome was really different, but still no successful compilation, as `clang` could not find the exported functions from the `Objective-C` file. It found the same function name with an `_` in front of the name.

As I'm no `C++` wizard I'm not sure what to try next.

Here the `Objective-C` code I tried:
```c
#import <stdlib.h>
#import <Foundation/Foundation.h>
#import <Cocoa/Cocoa.h>

// the include file just defines the functions in this file
#include "import_extern_obj-c.h"

void Export_Function() {    
    printf("Hello from Objective-C!\n");
}

int GetBitsPerPixel(int index) {
    NSScreen *screen = [NSScreen mainScreen];
    NSWindowDepth depth = [screen depth];

    return (int) NSBitsPerPixelFromDepth (depth);
}
```
I did the same test with the `Vlang` compiler utilizing the same `Objective-C` file and it compiled flawlessly. 
However, `Vlang` compiles to `C`, which can be used in conjunction with `Objective-C`...

**Reply by mertcandav** [reply]
Hi.

JuleC is currently in its very early stages and its interoperability with Objective-C and Objective-C++ has not been adequately tested. I've done a few tries myself, specific to your issue, and I think I've had a little of experience. Now I will tell you how I compile.

This is my Clang version: `Apple clang version 14.0.3 (clang-1403.0.22.14.1)`
I'm using latest commit of JuleC's master branch.

I used header file for declarations.
Here is my `objc.h` file:
```mm
#ifndef OBJC_H
#define OBJC_H

void Export_Function();
int GetBitsPerPixel(int index);

#endif
```

The Objective-C source code is your code.
Here is my `objc.mm` file:
```mm
#import <stdlib.h>
#import <Foundation/Foundation.h>
#import <Cocoa/Cocoa.h>

void Export_Function() {    
    printf("Hello from Objective-C!\n");
}

int GetBitsPerPixel(int index) {
    NSScreen *screen = [NSScreen mainScreen];
    NSWindowDepth depth = [screen depth];

    return (int) NSBitsPerPixelFromDepth (depth);
}
```

Finally, I wrote my Jule code and bind to my header file.
Here is my `main.jule` file:
```
use cpp "objc.h"

cpp fn Export_Function()
cpp fn GetBitsPerPixel(index: int): int

fn main() {
    cpp.Export_Function()
    let depth = cpp.GetBitsPerPixel(0)
    outln(depth)
}
```

I transpiled my Jule code to C++ first, and I compiled with Clang manually.
Here is my compile command: ``clang++ --std=c++14 -framework Foundation -framework Cocoa dist/ir.cpp objc.mm``

It's works fine for me. I hope it helps.

**Comment by sparky4pro** [comment]
Using your code worked, which is awesome!
Thank you so much.

Looking at the differences to what I did:
I used a `*.m` file extension instead of `*.mm` and because of this I had to avoid using `--std=c++14` which opened the can of worms I described earlier.

Using `--std=c++14` gave me the following error:
`error: invalid argument '--std=c++14' not allowed with 'Objective-C'`

Changing the file extension from `*.m` to `*.mm` did the trick, it compiled.

Thank you so much for helping me out.

The only thing that is left to do (on your side :D ) is to add something to add the `*.mm` file to the compilation process.
Maybe:
`link cpp "objc.mm"`
or 
`bind cpp "objc.mm"`
or, if jule can parse the include string:
`use cpp "objc.mm"`
and because it has the `*.mm` file extension, to add it to the compile call (clang) instead of adding an include statement into `ir.cpp` file, like with normal `*.h` files.

Again, thank you so much for helping me out.

This is so awesome!

EDIT:
Forgot to mention that `-framework Foundation -framework Cocoa` needs to be added as well.
Maybe allow adding: 
`flag -framework Foundation -framework Cocoa` 
or
`flag cpp -framework Foundation -framework Cocoa` 
to the `jule` code could do the trick.

**Reply by mertcandav** [reply]
You're welcome, thanks for your interest.

JuleC is not currently supports header files for only declarations. For example, if you want to use a ``.cpp`` file for a ``.hpp`` file, you can do it with manual compile, otherwise the ``.hpp`` file must contain all the code. A number of design additions are in the plans to address this.

Here are examples of potential additions considered for this:
```
use cpp "objc.h" for "objc.mm"
```

```
cpp impl "objc.mm" for "objc.h"
```

I will also consider your suggestion about adding flags. If we add this similar to what you said, it will probably take place as a Jule directive be like ``//jule:cflag -framework Foundation``

**Reply by sparky4pro** [reply]
The potential additions sound fantastic, and a Jule flag would be fine as well.

While thinking of it... using:
`//jule:clink objc.mm`
would work as well.

Thank you for adapting Jule.

BTW: I have a minimalistic macOS Cocoa GUI library (Objective-C) that I already use for `Go` and `V`.
With your help I can add the usability to `Jule` as well. :D 
Thank you.

**Reply by mertcandav** [reply]
If your library is public, you can add it to the community projects list when it is created.

It's good to see you consider Jule.
Many thanks for your interest and contributions.

**Reply by sparky4pro** [reply]
My library is not public yet, but will be... too many changes happening.

It was originally written for `Go` a few years back and a few month ago I've discovered `V` and started using it.
To adapt the library to make it work better with `V`, I had many changes to do. 
Now the `Go` and the `V` versions of my (Objective-C) library are out of sync (the `V` version being better and up-to-date).

When I discovered `Jule` I liked it very much so, that I plan to make my library work with `Jule` even before I change the `Go` glue code to use the newer version of my library.

BTW: you mentioned a  community projects list...
Do you have set-up a Forum or Discord server for Jule? 
Or is the Github discussion the only channel to talk to other Jule users?

**Reply by mertcandav** [reply]
It is a motivating and pleasant feeling for us to see those who love and use Jule.
I wish you good luck and good work in your projects.

In the early stages, the Jule Discord community existed for a certain period of time, however, it was closed due to inactivity. It seems good to have a Discord community these days when the community is more vibrant. An invitation link will be added to the relevant places soon. Until that happens, the only communication channel for now is GitHub Discussions.
